File: python/open_guardrail/guards/refusal_quality.py
"""Check quality of AI refusals."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_HARSH: list[re.Pattern[str]] = [
    re.compile(
        r"\byou\s+should\s+be\s+ashamed\b",
        re.IGNORECASE,
    ),
    re.compile(r"\bhow\s+dare\s+you\b", re.IGNORECASE),
    re.compile(
        r"\bthat'?s\s+(?:a\s+)?"
        r"(?:stupid|dumb|idiotic)"
        r"\s+(?:question|request)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bI\s+refuse\s+to\s+(?:even\s+)?"
        r"(?:dignify|acknowledge)\b",
        re.IGNORECASE,
    ),
]

_VAGUE: list[re.Pattern[str]] = [
    re.compile(
        r"^I\s+can'?t\s+help\s+with\s+that\.?\s*$",
        re.IGNORECASE | re.MULTILINE,
    ),
    re.compile(
        r"^I'?m\s+(?:unable|not\s+able)"
        r"\s+to\s+(?:help|assist)\.?\s*$",
        re.IGNORECASE | re.MULTILINE,
    ),
    re.compile(
        r"^(?:Sorry|No),?\s+I\s+can'?t\.?\s*$",
        re.IGNORECASE | re.MULTILINE,
    ),
    re.compile(
        r"^I\s+(?:can'?t|won'?t)\s+do\s+that\.?\s*$",
        re.IGNORECASE | re.MULTILINE,
    ),
]

_NO_REASON: re.Pattern[str] = re.compile(
    r"\bI\s+(?:can'?t|won'?t|am\s+unable\s+to)\b"
    r"(?!.*\b(?:because|since|as|due\s+to"
    r"|the\s+reason|in\s+order\s+to"
    r"|however|instead)\b)",
    re.IGNORECASE,
)

_ILLEGAL_ALT: list[re.Pattern[str]] = [
    re.compile(
        r"\binstead\s*,?\s*(?:you\s+(?:could|might|can)"
        r"\s+)?(?:try|use|go\s+to)\s+(?:the\s+)?"
        r"(?:dark\s*web|tor|black\s*market)",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bthere\s+are\s+(?:other\s+)?"
        r"(?:illegal|underground)"
        r"\s+(?:ways|methods|channels)\b",
        re.IGNORECASE,
    ),
]


class _RefusalQuality:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "refusal-quality"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if any(p.search(text) for p in _HARSH):
            issues.append("harsh_refusal")
        if any(p.search(text) for p in _VAGUE):
            issues.append("vague_refusal")
        if _NO_REASON.search(text) and len(text) < 200:
            issues.append("no_explanation")
        if any(p.search(text) for p in _ILLEGAL_ALT):
            issues.append("illegal_alternative")

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="refusal-quality",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Poor refusal quality detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def refusal_quality(
    *, action: str = "warn"
) -> _RefusalQuality:
    return _RefusalQuality(action=action)

File: packages/guards/src/refusal-quality.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RefusalQualityOptions {
  action: 'block' | 'warn';
}

const HARSH_PATTERNS: RegExp[] = [
  /\byou\s+should\s+be\s+ashamed\b/i,
  /\bhow\s+dare\s+you\b/i,
  /\bthat'?s\s+(?:a\s+)?(?:stupid|dumb|idiotic)\s+(?:question|request)\b/i,
  /\bI\s+refuse\s+to\s+(?:even\s+)?(?:dignify|acknowledge)\b/i,
];

const VAGUE_PATTERNS: RegExp[] = [
  /^I\s+can'?t\s+help\s+with\s+that\.?\s*$/im,
  /^I'?m\s+(?:unable|not\s+able)\s+to\s+(?:help|assist)\.?\s*$/im,
  /^(?:Sorry|No),?\s+I\s+can'?t\.?\s*$/im,
  /^I\s+(?:can'?t|won'?t)\s+do\s+that\.?\s*$/im,
];

const NO_REASON_PATTERNS: RegExp[] = [
  /\bI\s+(?:can'?t|won'?t|am\s+unable\s+to)\b(?!.*\b(?:because|since|as|due\s+to|the\s+reason|in\s+order\s+to|however|instead)\b)/i,
];

const ILLEGAL_ALT_PATTERNS: RegExp[] = [
  /\binstead\s*,?\s*(?:you\s+(?:could|might|can)\s+)?(?:try|use|go\s+to)\s+(?:the\s+)?(?:dark\s*web|tor|black\s*market)/i,
  /\bthere\s+are\s+(?:other\s+)?(?:illegal|underground)\s+(?:ways|methods|channels)\b/i,
];

export function refusalQuality(options: RefusalQualityOptions): Guard {
  return {
    name: 'refusal-quality',
    version: '0.1.0',
    description: 'Checks quality of AI refusals',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (HARSH_PATTERNS.some((p) => p.test(text))) issues.push('harsh_refusal');
      if (VAGUE_PATTERNS.some((p) => p.test(text))) issues.push('vague_refusal');
      if (NO_REASON_PATTERNS.some((p) => p.test(text)) && text.length < 200) {
        issues.push('no_explanation');
      }
      if (ILLEGAL_ALT_PATTERNS.some((p) => p.test(text))) issues.push('illegal_alternative');

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'refusal-quality',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
